### ISSUE TITLE
Failure to create Redis client (no viable redis host found)

### ISSUE THREAD

**Issue by lthibault** [testground/testground#954]
#### Version information:
Using Testground commit `c1ce7c4b836218be88c531765908984db6d7ec99`.

macOS Catalina v10.15.4

Docker:
```
Client: Docker Engine - Community
 Version:           19.03.8
 API version:       1.40
 Go version:        go1.12.17
 Git commit:        afacb8b
 Built:             Wed Mar 11 01:21:11 2020
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.8
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.17
  Git commit:       afacb8b
  Built:            Wed Mar 11 01:29:16 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.13
  GitCommit:        7ad184331fa3e55e52b890ea95e65ba581ae3429
 runc:
  Version:          1.0.0-rc10
  GitCommit:        dc9208a3303feef5b3839f4323d9beb36df0a9dd
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

#### Description:

After successfully running the example test plans using the `docker:go` builder and the `local:docker` runner, I wrote a small test case that uses the `sync` package from the sdk.  Importantly, I am testing a _private_  github repo, so after failing to find a suitable means of adding private deploy keys to the docker containers used by testground, I decided to try my luck with the `exec:go` builder and `local:exec` runner.

Here is the command I used:

```bash
testground run single --plan example --testcase sync --builder exec:go --runner local:exec --instances=2
```

and here is the output (truncated for brevity):

```
May  4 10:40:13.768844  INFO    using home directory: /Users/lthibault/testground
May  4 10:40:13.769012  INFO    no .env.toml found at /Users/lthibault/testground/.env.toml; running with defaults
May  4 10:40:13.769031  INFO    testground client initialized   {"addr": "localhost:8042"}
May  4 10:40:13.770539  INFO    using home directory: /Users/lthibault/testground
May  4 10:40:13.770605  INFO    no .env.toml found at /Users/lthibault/testground/.env.toml; running with defaults
May  4 10:40:13.770615  INFO    testground client initialized   {"addr": "localhost:8042"}
May  4 10:40:13.771064  INFO    test plan source at: /Users/lthibault/testground/plans/example

>>> Server output:

May  4 10:40:13.790966  INFO    performing build for groups     {"req_id": "a9bdd9a8", "plan": "example", "groups": ["single"], "builder": "exec:go"}
May  4 10:40:17.055034  INFO    build succeeded {"req_id": "a9bdd9a8", "plan": "example", "groups": ["single"], "builder": "exec:go", "artifact": "/Users/lthibault/testground/data/work/exec-go--example-defe629a76a8"}

>>> Result:

May  4 10:40:17.056337  INFO    generated build artifact        {"group": "single", "artifact": "/Users/lthibault/testground/data/work/exec-go--example-defe629a76a8"}

>>> Server output:

May  4 10:40:17.061313  INFO    performing healthcheck on runner        {"req_id": "456deef1"}
May  4 10:40:17.113120  INFO    healthcheck: ok {"req_id": "456deef1"}
May  4 10:40:17.114303  INFO    starting test case instance     {"req_id": "456deef1", "plan": "example", "group": "single", "number": 0, "total": 1}
May  4 10:40:17.117940  INFO    starting test case instance     {"req_id": "456deef1", "plan": "example", "group": "single", "number": 1, "total": 2}
May  4 10:40:17.657870  INFO    0.5439s    MESSAGE << instance   1 >> registering default http handler at: http://[::]:6060/ (pprof: http://[::]:6060/debug/pprof/)       {"req_id": "456deef1"}
May  4 10:40:17.658138  INFO    0.5440s      START << instance   1 >> {"plan":"example","case":"sync","run":"ccab27dc5c0f","outputs_path":"/Users/lthibault/testground/data/outputs/local_exec/example/ccab27dc5c0f/single/0","instances":2,"group":"single","group_instances":2,"network":"127.1.0.0/16","start_time":"0001-01-01T00:00:00Z"}        {"req_id": "456deef1"}
May  4 10:40:17.658286  INFO    0.5441s    MESSAGE << instance   2 >> registering default http handler at: http://[::]:52643/ (pprof: http://[::]:52643/debug/pprof/)     {"req_id": "456deef1"}
May  4 10:40:17.658455  INFO    0.5442s      START << instance   2 >> {"plan":"example","case":"sync","run":"ccab27dc5c0f","outputs_path":"/Users/lthibault/testground/data/outputs/local_exec/example/ccab27dc5c0f/single/1","instances":2,"group":"single","group_instances":2,"network":"127.1.0.0/16","start_time":"0001-01-01T00:00:00Z"}        {"req_id": "456deef1"}
May  4 10:40:17.666746  INFO    0.5528s      CRASH << instance   2 >> failed to create redis client: no viable redis host foundgoroutine 1 [running]:
runtime/debug.Stack(0x1629757, 0x2, 0xc00014d9d8)
        /usr/local/go/src/runtime/debug/stack.go:24 +0x9d
github.com/testground/sdk-go/runtime.(*RunEnv).RecordCrash(0xc00013e000, 0x1589d80, 0xc000198440)
        /Users/lthibault/go/pkg/mod/github.com/testground/sdk-go@v0.2.0/runtime/runenv_events.go:144 +0xa2
github.com/testground/sdk-go/runtime.Invoke.func2(0xc00013e000)
        /Users/lthibault/go/pkg/mod/github.com/testground/sdk-go@v0.2.0/runtime/runner.go:87 +0x60
panic(0x1589d80, 0xc000198440)
        /usr/local/go/src/runtime/panic.go:967 +0x15d
github.com/testground/sdk-go/sync.MustBoundClient(...)
        /Users/lthibault/go/pkg/mod/github.com/testground/sdk-go@v0.2.0/sync/client.go:79
main.ExampleSync(0xc00013e000, 0x0, 0x0)
        /Users/lthibault/testground/data/work/requests/a9bdd9a8/plan/sync.go:29 +0xaee
main.run(0xc00013e000, 0x16521e8, 0xc00013c720)
        /Users/lthibault/testground/data/work/requests/a9bdd9a8/plan/main.go:27 +0x122
github.com/testground/sdk-go/runtime.Invoke(0x16524f8)
        /Users/lthibault/go/pkg/mod/github.com/testground/sdk-go@v0.2.0/runtime/runner.go:95 +0x2b2
main.main()
        /Users/lthibault/testground/data/work/requests/a9bdd9a8/plan/main.go:11 +0x2d
        {"req_id": "456deef1"}
May  4 10:40:17.667428  INFO    0.5537s      ERROR << instance   1 >> failed to create redis client: no viable redis host found   {"req_id": "456deef1"}
May  4 10:40:17.667506  INFO    0.5538s      ERROR << instance   1 >> goroutine 1 [running]:      {"req_id": "456deef1"}
```

After confirming that Redis was, indeed, running in Docker, I attempted to set the `REDIS_HOST` and `REDIS_PORT` environment variables to `127.0.0.1` and `6379`, respectively.  N.B.:  I set these environment variables at the OS level, not in Docker.

Lastly, I attempted to run the `sync` test case from the `example` plan using the `exec:go` builider and `local:exec` runner.  As expected, this produced the same error as above.

I'm at your disposal for further diagnostics, should these prove helpful.

Kind regards,
Louis

**Comment by raulk** [comment]
@lthibault Thanks for reporting this. It's pretty weird that the healthcheck was successful, yet we fail to connect to Redis:

```
May  4 10:40:17.113120  INFO    healthcheck: ok {"req_id": "456deef1"}
```

This could indicate that the `testground-redis` container is running, but it may not be accessible for some reason.

Could you try these three things?

1. Run `docker ps` and paste the output -- we're expecting to see a `testground-redis` container started and with port 6379 exposed onto the host.
2. Run `docker inspect testground-redis`, and paste the output.
3. From your local console, telnet into Redis: `telnet localhost 6379`, and send a `PING` command.

**Comment by raulk** [comment]
> I am testing a private github repo

This is a use case we hadn't considered. You can import test plans from private repos, but when test plans depend on private repos themselves -- that's trickier! I've opened a feature request to track this: https://github.com/testground/testground/issues/955.

**Comment by lthibault** [comment]
@raulk My pleasure -- thank you for the prompt reply!

Here's the output of `docker ps`:

```$ docker ps
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                    NAMES
f07fd07641b6        influxdb:1.8        "/entrypoint.sh infl…"   51 minutes ago      Up 51 minutes       0.0.0.0:8086->8086/tcp   testground-influxdb
2c8815e6e0e8        redis               "docker-entrypoint.s…"   51 minutes ago      Up 51 minutes       0.0.0.0:6379->6379/tcp   testground-redis
49fcfd8219b2        bitnami/grafana     "/run.sh"                51 minutes ago      Up 51 minutes       0.0.0.0:3000->3000/tcp   testground-grafana
```

Here's the output of `docker inspect testground-redis`:

```
[
    {
        "Id": "2c8815e6e0e8c72c2caf4a4fd586e8ad4ad1a8fc608aef2fd3e56b9f464f1aa5",
        "Created": "2020-05-04T10:25:29.8880409Z",
        "Path": "docker-entrypoint.sh",
        "Args": [
            "--save",
            "\"\"",
            "--appendonly",
            "no",
            "--maxclients",
            "120000"
        ],
        "State": {
            "Status": "running",
            "Running": true,
            "Paused": false,
            "Restarting": false,
            "OOMKilled": false,
            "Dead": false,
            "Pid": 28033,
            "ExitCode": 0,
            "Error": "",
            "StartedAt": "2020-05-04T10:25:32.3872758Z",
            "FinishedAt": "0001-01-01T00:00:00Z"
        },
        "Image": "sha256:f9b9909726890b00d2098081642edf32e5211b7ab53563929a47f250bcdc1d7c",
        "ResolvConfPath": "/var/lib/docker/containers/2c8815e6e0e8c72c2caf4a4fd586e8ad4ad1a8fc608aef2fd3e56b9f464f1aa5/resolv.conf",
        "HostnamePath": "/var/lib/docker/containers/2c8815e6e0e8c72c2caf4a4fd586e8ad4ad1a8fc608aef2fd3e56b9f464f1aa5/hostname",
        "HostsPath": "/var/lib/docker/containers/2c8815e6e0e8c72c2caf4a4fd586e8ad4ad1a8fc608aef2fd3e56b9f464f1aa5/hosts",
        "LogPath": "/var/lib/docker/containers/2c8815e6e0e8c72c2caf4a4fd586e8ad4ad1a8fc608aef2fd3e56b9f464f1aa5/2c8815e6e0e8c72c2caf4a4fd586e8ad4ad1a8fc608aef2fd3e56b9f464f1aa5-json.log",
        "Name": "/testground-redis",
        "RestartCount": 0,
        "Driver": "overlay2",
        "Platform": "linux",
        "MountLabel": "",
        "ProcessLabel": "",
        "AppArmorProfile": "",
        "ExecIDs": null,
        "HostConfig": {
            "Binds": null,
            "ContainerIDFile": "",
            "LogConfig": {
                "Type": "json-file",
                "Config": {}
            },
            "NetworkMode": "testground-control",
            "PortBindings": {
                "6379/tcp": [
                    {
                        "HostIp": "",
                        "HostPort": "6379"
                    }
                ]
            },
            "RestartPolicy": {
                "Name": "unless-stopped",
                "MaximumRetryCount": 0
            },
            "AutoRemove": false,
            "VolumeDriver": "",
            "VolumesFrom": null,
            "CapAdd": null,
            "CapDrop": null,
            "Capabilities": null,
            "Dns": null,
            "DnsOptions": null,
            "DnsSearch": null,
            "ExtraHosts": null,
            "GroupAdd": null,
            "IpcMode": "private",
            "Cgroup": "",
            "Links": null,
            "OomScoreAdj": 0,
            "PidMode": "",
            "Privileged": false,
            "PublishAllPorts": false,
            "ReadonlyRootfs": false,
            "SecurityOpt": null,
            "UTSMode": "",
            "UsernsMode": "",
            "ShmSize": 67108864,
            "Sysctls": {
                "net.core.somaxconn": "150000",
                "net.netfilter.nf_conntrack_max": "120000"
            },
            "Runtime": "runc",
            "ConsoleSize": [
                0,
                0
            ],
            "Isolation": "",
            "CpuShares": 0,
            "Memory": 0,
            "NanoCpus": 0,
            "CgroupParent": "",
            "BlkioWeight": 0,
            "BlkioWeightDevice": null,
            "BlkioDeviceReadBps": null,
            "BlkioDeviceWriteBps": null,
            "BlkioDeviceReadIOps": null,
            "BlkioDeviceWriteIOps": null,
            "CpuPeriod": 0,
            "CpuQuota": 0,
            "CpuRealtimePeriod": 0,
            "CpuRealtimeRuntime": 0,
            "CpusetCpus": "",
            "CpusetMems": "",
            "Devices": null,
            "DeviceCgroupRules": null,
            "DeviceRequests": null,
            "KernelMemory": 0,
            "KernelMemoryTCP": 0,
            "MemoryReservation": 0,
            "MemorySwap": 0,
            "MemorySwappiness": null,
            "OomKillDisable": false,
            "PidsLimit": null,
            "Ulimits": [
                {
                    "Name": "nofile",
                    "Hard": 1048576,
                    "Soft": 1048576
                }
            ],
            "CpuCount": 0,
            "CpuPercent": 0,
            "IOMaximumIOps": 0,
            "IOMaximumBandwidth": 0,
            "MaskedPaths": [
                "/proc/asound",
                "/proc/acpi",
                "/proc/kcore",
                "/proc/keys",
                "/proc/latency_stats",
                "/proc/timer_list",
                "/proc/timer_stats",
                "/proc/sched_debug",
                "/proc/scsi",
                "/sys/firmware"
            ],
            "ReadonlyPaths": [
                "/proc/bus",
                "/proc/fs",
                "/proc/irq",
                "/proc/sys",
                "/proc/sysrq-trigger"
            ]
        },
        "GraphDriver": {
            "Data": {
                "LowerDir": "/var/lib/docker/overlay2/a398efef5d21d8dd9e21390356bddbe39269f69a1752c6b32f7b8c205b9ed2dc-init/diff:/var/lib/docker/overlay2/38120696fc2e432e61fc97314d5d83d2fb318d4c469c659bf4346db892774bd6/diff:/var/lib/docker/overlay2/3134e31012ec6e7436e1478ea754a75134da7fc648706792036358812ddb688c/diff:/var/lib/docker/overlay2/a4c173b77edc4e59596b028d439478422e428f85fe6875f26f7b6d3b99ebb026/diff:/var/lib/docker/overlay2/592a5345340c5f80ff2aff2d0bc4ae629b3df726a154598939252db91687c0a1/diff:/var/lib/docker/overlay2/eb6bc621c8e471692a6041c8dc77b933ddb844680488ee04ad82a0d1a94b449a/diff:/var/lib/docker/overlay2/62e15126557c022eb62777019a45571aacb333913cc6d6c58ee1d5b18c514158/diff",
                "MergedDir": "/var/lib/docker/overlay2/a398efef5d21d8dd9e21390356bddbe39269f69a1752c6b32f7b8c205b9ed2dc/merged",
                "UpperDir": "/var/lib/docker/overlay2/a398efef5d21d8dd9e21390356bddbe39269f69a1752c6b32f7b8c205b9ed2dc/diff",
                "WorkDir": "/var/lib/docker/overlay2/a398efef5d21d8dd9e21390356bddbe39269f69a1752c6b32f7b8c205b9ed2dc/work"
            },
            "Name": "overlay2"
        },
        "Mounts": [
            {
                "Type": "volume",
                "Name": "60d27d895ade85d352388c54e5186cddbb51ea699e17d493e6e6729ea731e51e",
                "Source": "/var/lib/docker/volumes/60d27d895ade85d352388c54e5186cddbb51ea699e17d493e6e6729ea731e51e/_data",
                "Destination": "/data",
                "Driver": "local",
                "Mode": "",
                "RW": true,
                "Propagation": ""
            }
        ],
        "Config": {
            "Hostname": "2c8815e6e0e8",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "6379/tcp": {}
            },
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": [
                "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
                "GOSU_VERSION=1.12",
                "REDIS_VERSION=6.0.1",
                "REDIS_DOWNLOAD_URL=http://download.redis.io/releases/redis-6.0.1.tar.gz",
                "REDIS_DOWNLOAD_SHA=b8756e430479edc162ba9c44dc89ac394316cd482f2dc6b91bcd5fe12593f273"
            ],
            "Cmd": [
                "--save",
                "\"\"",
                "--appendonly",
                "no",
                "--maxclients",
                "120000"
            ],
            "Image": "library/redis",
            "Volumes": {
                "/data": {}
            },
            "WorkingDir": "/data",
            "Entrypoint": [
                "docker-entrypoint.sh"
            ],
            "OnBuild": null,
            "Labels": {}
        },
        "NetworkSettings": {
            "Bridge": "",
            "SandboxID": "4b9b73a72320996a77996ba253515d9b28f934d5af1dc7009ebf3b5c2f0f69ae",
            "HairpinMode": false,
            "LinkLocalIPv6Address": "",
            "LinkLocalIPv6PrefixLen": 0,
            "Ports": {
                "6379/tcp": [
                    {
                        "HostIp": "0.0.0.0",
                        "HostPort": "6379"
                    }
                ]
            },
            "SandboxKey": "/var/run/docker/netns/4b9b73a72320",
            "SecondaryIPAddresses": null,
            "SecondaryIPv6Addresses": null,
            "EndpointID": "",
            "Gateway": "",
            "GlobalIPv6Address": "",
            "GlobalIPv6PrefixLen": 0,
            "IPAddress": "",
            "IPPrefixLen": 0,
            "IPv6Gateway": "",
            "MacAddress": "",
            "Networks": {
                "testground-control": {
                    "IPAMConfig": null,
                    "Links": null,
                    "Aliases": [
                        "2c8815e6e0e8"
                    ],
                    "NetworkID": "8cf09ef9c3d80d7d555f9b3b67e9bc928d3484668b9f9cb6472cf09f955f2ba3",
                    "EndpointID": "d6fc9d9fd572963a39971e210091296d2f5c2423c70cffba5b41a05f302cbcff",
                    "Gateway": "192.18.0.1",
                    "IPAddress": "192.18.0.3",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:c0:12:00:03",
                    "DriverOpts": null
                }
            }
        }
    }
]
```
And lastly, the telnet diagnostic seems to work fine:

```
 $ telnet localhost 6379  # I also tried with ipv4 using 127.0.0.1.  also works.
Trying ::1...
Connected to localhost.
Escape character is '^]'.
ping
-MISCONF Redis is configured to save RDB snapshots, but it is currently not able to persist on disk. Commands that may modify the data set are disabled, because this instance is configured to report errors during writes if RDB snapshotting fails (stop-writes-on-bgsave-error option). Please check the Redis logs for details about the RDB error.
```

>This is a use case we hadn't considered. You can import test plans from private repos, but when test plans depend on private repos themselves -- that's trickier! I've opened a feature request to track this: #955.

Great, thanks!  In practice, this seems to work fine with the local builder/runner - as long as the host OS is able to `go get` the private repo, testground runs smoothly.  From where I'm standing, the issue appears to be at the interface of Redis and the `sync.Client`.

Thanks again for your time, and please let me know if I can do anything else to get this resolved.

**Comment by raulk** [comment]
@lthibault I've adjusted the CLI parameters we start Redis with in #956. Could you please check out the `fix/redis-rdb` branch and try running that?

Could you do the following and report back?

1. `docker rm -f testground-redis` to delete the redis container.
2. `go build .` to build testground.
3. stop the daemon if it's running.
4. start the newly built daemon again.
5. `testground healthcheck --runner local:docker --fix` should recreate the redis container (we anyway do this automatically on runs).
6. try to run the test plan again

Basically what was happening is that Redis was refusing to service requests because it couldn't write to its RDB location, but RDB snapshots are supposed to be disabled anyway.

**Comment by lthibault** [comment]
@raulk Sure thing.  Do I need to rebuild the project (i.e. `make install`) for this?

**Comment by raulk** [comment]
@lthibault nope, just building testground via `go build .` from the root of this repo should be enough ;-)

**Comment by lthibault** [comment]
@raulk 

I'm getting the same output as before, including the `-MISCONF` message in Redis.  It seems `testground healthcheck --runner local:docker --fix` didn't rebuild the container?

I'm going to try deleting the corresponding Redis image, `docker rm` everything, and try again.

**Comment by lthibault** [comment]
@raulk No joy.  I'm getting exactly the same output as before.

Here's my sanity checklist:

- re-built the binary with `go build .`
- terminated the daemon, and ran `./testground daemon` using the newly built binary
- killed and removed redis container
- used `docker rmi` to remove the redis image (you never know ...)
- ran `./testground healthcheck --runner local:docker --fix` with the newly built binary
- ran `./testground run single --plan example --testcase sync --builder exec:go --runner local:ex
ec --instances=2` with the newly built binary
- ran telnet/ping against redis.  Got the same `-MISCONF` message.

For good measure, I then ran:

```bash
./testground run single --plan example --testcase sync --builder docker:go --runner local:docker --instances=2
```

Which worked (as expected).

Thoughts?

**Comment by raulk** [comment]
@lthibault I wonder if there's another Redis process on your machine. Could you call: `testground terminate --runner local:docker`? You should no longer see a `testground-redis` container in `docker ps`.

Then can you try the telnet again? If it connects, there's another Redis process on your machine that may be the culprit...

**Comment by lthibault** [comment]
>I wonder if there's another Redis process on your machine.

Ha! Bingo!  I forgot I had a Redis instance running as a homebrew service.

```
$ brew services list
Name       Status  User      Plist
ipfs       started lthibault /Users/lthibault/Library/LaunchAgents/homebrew.mxcl.ipfs.plist
redis      started lthibault /Users/lthibault/Library/LaunchAgents/homebrew.mxcl.redis.plist
```

I've stopped the service and re-run `./testground run single --plan example --testcase sync --builder exec:go --runner local:exec --instances=2`, which now works like a charm.

I wonder if there's a way to detect this issue during the health check?  Or perhaps better yet, we could run Testground's Redis container on a non-standard port?

Thanks again for your help.

P.S.:  is there any way I could have access to the Testground Git Book?  I understand it's an early version and subject to sudden change, but I'm quite eager to use and contribute to Testground.

**Comment by raulk** [comment]
@lthibault awesome! 🎉 Glad we sorted this out. I like your suggestion -- if the container is not running, we should detect if _something_ is already running on port 6379. Mind opening an issue to track that? PRs are totally welcome :-)

The Gitbook for now is published at https://protocol-labs.gitbook.io/testground/ -- this is actively WIP and we're fleshing it out these days, so some sections are messy/empty.

**Comment by lthibault** [comment]
@raulk Awesome indeed! 

>if the container is not running, we should detect if something is already running on port 6379.

Sounds good.  I'll open the issue and take a stab at it (probably sometime this week).

Thanks for the link!
(Closing this issue since it's now working on my end)
